bids/layout/layout.py

```diff
@@ -114,7 +114,10 @@ def __init__(self, root, validate=True, index_associated=True,
                                      "dataset_description.json." % k)
 
         # Determine which subdirectories to exclude from indexing
-        excludes = {"code", "stimuli", "sourcedata", "models", "derivatives"}
+        # A set is extended with non-BIDS-defined common suspects
+        # which are not expected to carry any data of interest
+        excludes = {"code", "stimuli", "sourcedata", "models", "derivatives",
+                    ".git", ".datalad", ".cache", ".workdir", "workdir"}
```

Comment:
I agree `.git` should never be considered. I'm not sure `.bidsignore` reading is implemented.  That would be a valuable contribution. But I also think sensible hardcoded defaults (such as these) make sense.

Reply:
Also, please update the docstring

Reply:
If we're going to specify sensible defaults, I'd be inclined to disallow all `.*` files except `.bidsignore`.

Reply:
The reason these particular values are hardcoded is that they're explicitly mentioned in the spec as places people should put stuff, but we don't want to index them by default. The other directories (e.g., .git) are not part of the spec, so I don't want to give them special handling. I we start adding things like `workdir/`, there's no limit to what people might want to add.

I agree that `.bidsignore` support is the right way to handle this. Pybids currently passes (in `kwargs`) an `exclude` argument onto grabbit, which one can use to pass arbitrary regexes that (I believe) are matched against every file. So probably the easiest way to implement `.bidsignore` support would be to add entries in `.bidsignore` to the `exclude` list. There may be unintended consequences, however, so this should be investigated.

Reply:
> If we're going to specify sensible defaults, I'd be inclined to disallow all `.*` files except `.bidsignore`.

Just curious -- what for to allow/consider .bidsignore?  if any function would be needing to treat this this file, it would just test for its presence and open it. It doesn't need it to be a part of the Layout schema/pool of files, since there is nothing to query for in its name.

Reply:
> The other directories (e.g., .git) are not part of the spec, so I don't want to give them special handling. I we start adding things like `workdir/`, there's no limit to what people might want to add.
> 
> I agree that `.bidsignore` support is the right way to handle this.

I agree too! Support of .bidsignore would allow for all those `workdir` etc. But also I agree with @adelavega -- I do not think in any BIDS spec we would/should allow/be interested in files/directories starting with `.` for the purpose of paths matching etc.
As I have pointed above, `.bidsignore` is a special file which is not part of the path layout template matching and actually not a part of the BIDS specification itself!  Its existence was instigated within bids-validator, so it somewhat of a tooling level file.  So it should IMHO be ok to ignore all `.` files/dirs in general for pattern matching. I think it might even be worth to "hardcode" such consideration in BIDS specification itself? e.g. 

"Note on BIDS dataset tools support/development: No directories or files in BIDS are generally named with leading `.` and thus should be ignored for the purpose of discovery of relevant data files. A `.bidsignore` file can be used by dataset users to inform tools of additional directories/files which should also be ignored by the tools.  The format of `.bidsignore` file is .."

Reply:
Agreed—let's add "^\." to the list of default excludes (if I'm not mistaken, they're regexes by default) and remove the other entries. We can add `.bidsignore` support in a separate PR, unless someone wants to take a pass at it here.

I don't have strong feelings about whether or not the spec should mention `.` and `.bidsignore`. I guess I lean towards not for the time being; to my mind the place to put that would be some kind of developer guide or appendix.